### PR TITLE
A2ha env setup a2ha.rb integration

### PR DESCRIPTION
**Pull request by jayvikramsharma1**
### :nut_and_bolt: Description: What code changed, and why?
A2HA env setup in automate code, installation of automate-backend-development, integration of a2ha.rb file generation logic, a2ha deploy command mapping with automate deploy command mapping

### :chains: Related Resources
https://app.zenhub.com/workspaces/the-delta-quadrant-5fc7c4c3922dc10021323fa6/issues/chef/automate/5434

### :+1: Definition of Done
Written Codes, Tested codes locally, No test case is breaking,

### :athletic_shoe: How to Build and Test the Change

rebuild components/automate-deployment/
rebuild components/automate-cli/

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: components/automate-deployment/pkg/target/local_target.go [hunk 1]
```diff
@@ -1520,3 +1520,22 @@ func (t *LocalTarget) IPs() []net.IP {
...
 }
+
+func (t *LocalTarget) InstallAutomateBackendDeployment(ctx context.Context, m manifest.ReleaseManifest) error {
+	//var pkg habpkg.HabPkg = habpkg.New("chef", "automate-backend-deployment")
+	pkg := manifest.InstallableFromManifest(m, "automate-backend-deployment")
+	if pkg == nil {
+		logrus.Info("(HA) unable to find automate-backend-deployment package in manifest")
+		return errors.New("automate-backend-deployment (HA) was not found in the manifest")
+	}
+	output, err := t.InstallPackage(ctx, pkg, "dev")
```
Comment (punitmundra) [inline components/automate-deployment/pkg/target/local_target.go:1531]: we can change dev -> c.GetV1().GetSvc().GetChannel().GetValue()
once automate will release next week

File: components/automate-cli/cmd/chef-automate/generate-a2harb.go [hunk 2]
```diff
@@ -69,3 +81,186 @@ func renderSettingsToA2HARBFile(templateName string, data interface{}) string {
...
+		return errors.New("Invalid or empty aws postgresql_ebs_volume_type")
+	} else {
+		return nil
+	}
+}
+
+func validateIPs(exitingInfra ExistingInfraConfigToml) error {
+	for _, element := range exitingInfra.ExistingInfra.Config.AutomateIps {
+		if checkIPAddress(element) != nil {
+			return errors.New("Automate Ip " + element + "is not valid")
```
Comment (punitmundra) [inline components/automate-cli/cmd/chef-automate/generate-a2harb.go:212]: in case of public ip missing ....add err msg .....use private ip ...in public ip place

**Review by punitmundra — CHANGES_REQUESTED** [review]
just test the full deployment ...once the testing is done...will approve the changes

File: components/automate-cli/cmd/chef-automate/generate-a2harb.go [hunk 3]
```diff
@@ -69,3 +81,186 @@ func renderSettingsToA2HARBFile(templateName string, data interface{}) string {
...
+	} else if len(awsConfig.Aws.Config.ElasticsearchEbsVolumeSize) < 1 {
+		return errors.New("Invalid or empty aws elasticsearch_ebs_volume_size")
+	} else if len(awsConfig.Aws.Config.ElasticsearchEbsVolumeType) < 1 {
+		return errors.New("Invalid or empty aws elasticsearch_ebs_volume_type")
+	} else if len(awsConfig.Aws.Config.PostgresqlEbsVolumeIops) < 1 {
+		return errors.New("Invalid or empty aws postgresql_ebs_volume_iops")
+	} else if len(awsConfig.Aws.Config.PostgresqlEbsVolumeSize) < 1 {
+		return errors.New("Invalid or empty aws postgresql_ebs_volume_size")
+	} else if len(awsConfig.Aws.Config.PostgresqlEbsVolumeType) < 1 {
+		return errors.New("Invalid or empty aws postgresql_ebs_volume_type")
```
Comment (jayvikramsharma1) [inline components/automate-cli/cmd/chef-automate/generate-a2harb.go:203]: return of error list to show all errors at one go

File: components/automate-cli/cmd/chef-automate/generate-a2harb.go [hunk 4]
```diff
@@ -69,3 +81,186 @@ func renderSettingsToA2HARBFile(templateName string, data interface{}) string {
...
+	} else if len(exitingInfra.ExistingInfra.Config.ChefServerPrivateIps) < 1 {
+		return errors.New("Invalid or empty chef_server_private_ips")
+	} else if len(exitingInfra.ExistingInfra.Config.ElasticsearchIps) < 1 {
+		return errors.New("Invalid or empty elasticsearch_ips")
+	} else if len(exitingInfra.ExistingInfra.Config.ElasticsearchPrivateIps) < 1 {
+		return errors.New("Invalid or empty elasticsearch_private_ips")
+	} else if len(exitingInfra.ExistingInfra.Config.PostgresqlIps) < 1 {
+		return errors.New("Invalid or empty postgresql_ips")
+	} else if len(exitingInfra.ExistingInfra.Config.PostgresqlPrivateIps) < 1 {
+		return errors.New("Invalid or empty postgresql_private_ips")
```
Comment (jayvikramsharma1) [inline components/automate-cli/cmd/chef-automate/generate-a2harb.go:155]: return of error list to show all errors at one go

File: components/automate-cli/cmd/chef-automate/generate-a2harb.go [hunk 5]
```diff
@@ -69,3 +81,186 @@ func renderSettingsToA2HARBFile(templateName string, data interface{}) string {
...
+	} else if len(awsConfig.Architecture.ConfigInitials.BackupMount) < 1 {
+		return errors.New("Invalid or empty backup_mount")
+	} else if len(awsConfig.Automate.Config.InstanceCount) < 1 {
+		return errors.New("Invalid or empty automate instance_count")
+	} else if len(awsConfig.ChefServer.Config.InstanceCount) < 1 {
+		return errors.New("Invalid or empty chef-server instance_count")
+	} else if len(awsConfig.Elasticsearch.Config.InstanceCount) < 1 {
+		return errors.New("Invalid or empty elastic-search instance_count")
+	} else if len(awsConfig.Postgresql.Config.InstanceCount) < 1 {
+		return errors.New("Invalid or empty postgres-sql instance_count")
```
Comment (jayvikramsharma1) [inline components/automate-cli/cmd/chef-automate/generate-a2harb.go:109]: return of error list to show all errors at one go

File: components/automate-cli/cmd/chef-automate/deploy_ha.go [hunk 6]
```diff
@@ -3,18 +3,68 @@
...
+		}
+	} else if checkIfFileExist(initConfigHabA2HAPathFlag.a2haDirPath + "a2ha.rb") {
+		return true, HA_MODE
 	} else {
+		return false, AUTOMATE
+	}
+
+}
+
+func deployA2HA() error {
```
Comment (jayvikramsharma1) [inline components/automate-cli/cmd/chef-automate/deploy_ha.go:41]: rename function to something like execute-deployment or triggerAutomateCTL

File: components/automate-cli/cmd/chef-automate/deploy_ha.go [hunk 7]
```diff
@@ -3,18 +3,68 @@
...
 	initConfigHAPath := initConfigHAPathFlags.path
-	config, err := ptoml.LoadFile(initConfigHAPath)
-	if err != nil {
-		return false
+	if len(configPath) > 0 {
+		initConfigHAPath = configPath[0]
 	}
-	if config.Get("architecture.existing_infra") != nil || config.Get("architecture.aws") != nil {
-		return true
+	if checkIfFileExist(initConfigHAPath) {
```
Comment (jayvikramsharma1) [inline components/automate-cli/cmd/chef-automate/deploy_ha.go:20]: Use some of design patterns to make code more manageable

**Comment by jayvikramsharma1** [comment]
@kalroy @vivek-yadav I have merged all my mapping tickets in this PR please have a quick review on it, we will be discussing more in our today 7:00 review discussion meeting

File: components/automate-cli/cmd/chef-automate/AutomateClusterCtlUtils.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
+package main
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/AutomateClusterCtlUtils.go:1]: why is the file name starting with capital letter?

File: components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+type awsDeployment struct{}
+
+func newAwsDeployemnt() *awsDeployment {
+	return &awsDeployment{}
+}
+
+func (a *awsDeployment) doDeployWork(args []string) error {
+	if isA2HARBFileExist() {
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go:14]: why do we need to check this again, if we had done this check when we did it in getDeployer->getModeFromConfig->checkIfFileExist

File: components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+	manager deployManager
+}
+
+func newDeployDirector(m deployManager) *deployDirector {
+	return &deployDirector{
+		manager: m,
+	}
+}
+
+func (d *deployDirector) setDeployManager(m deployManager) {
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go:13]: Why is this needed?

File: components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+	d.manager = m
+}
+
+func (d *deployDirector) executeDeployemnt(args []string) error {
+	return d.manager.doDeployWork(args)
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go:18]: Can we not call `manager.doDeployWorks` where `executeDeployment` is called from. As manager is already an interface var. And is available in `runDeployCmd`.    This whole layer will not be needed, as `getDeployer` is working as Factory Method.

File: components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+func (d *deployDirector) executeProvisionInfra(args []string) error {
+	return d.manager.doProvisionJob(args)
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go:22]: Same as above, we can directly call `manager.doProvisionJob` instead of `executeProvisionInfra` in `runProvisionInfraCmd` as delpoyManager is already available there.

File: components/automate-cli/cmd/chef-automate/automateDeployExecutor.go [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+	doProvisionJob(args []string) error
+}
+
+func getDeployer(args []string) deployManager {
+	deployerType := getModeFromConfig(args)
+	if deployerType == AWS_MODE {
+		return &awsDeployment{}
+	}
+	if deployerType == EXISTING_INFRA_MODE {
+		return &existingInfra{}
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateDeployExecutor.go:14]: call this `newExistingInfa` instead of direct instance creation.

File: components/automate-cli/cmd/chef-automate/automateDeployExecutor.go [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+type deployManager interface {
+	doDeployWork(args []string) error
+	doProvisionJob(args []string) error
+}
+
+func getDeployer(args []string) deployManager {
+	deployerType := getModeFromConfig(args)
+	if deployerType == AWS_MODE {
+		return &awsDeployment{}
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateDeployExecutor.go:11]: call this: `newAwsDeployemnt` instead of direct instance creation.

File: components/automate-cli/cmd/chef-automate/automateDeployExecutor.go [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+	}
+	if deployerType == EXISTING_INFRA_MODE {
+		return &existingInfra{}
+	}
+	if deployerType == HA_MODE {
+		return &haWithoutConfig{}
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateDeployExecutor.go:17]: same here, use this `newHaWithoutConfig`

File: components/automate-cli/cmd/chef-automate/secrets-ha.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+	Args: cobra.RangeArgs(0, 2),
+	RunE: runSecretsConfigCmd,
+}
+
+func runSecretsConfigCmd(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		writer.Print("please refer \n" + secretsHelpDocs)
+		return nil
+	}
+	return executeAutomateClusterCtlCommand("secrets", args, secretsHelpDocs)
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/secrets-ha.go:30]: Not sure, is `isA2HARBFileExist` check not needed in this case ?

File: components/automate-cli/cmd/chef-automate/workspace-ha.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+	Args: cobra.RangeArgs(0, 2),
+	RunE: runWorkspaceCmd,
+}
+
+func runWorkspaceCmd(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		writer.Print("please refer \n" + workspaceCommandHelpDocs)
+		return nil
+	}
+	return executeAutomateClusterCtlCommand("workspace", args, workspaceCommandHelpDocs)
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/workspace-ha.go:30]: Not sure, is 'isA2HARBFileExist' not needed here?

File: components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go [hunk 18]
```diff
@@ -0,0 +1,23 @@
+package main
+
+type deployDirector struct {
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go:3]: Builder pattern is used when Factory Pattern becomes too complex. or Object need to lot of steps to create. I feel Simple Factory pattern should be fine for us here.

File: components/automate-cli/cmd/chef-automate/automateConfigUtils.go [hunk 19]
```diff
@@ -0,0 +1,28 @@
+package main
+
+import ptoml "github.com/pelletier/go-toml"
+
+func getModeFromConfig(configPath []string) string {
+	initConfigHAPath := initConfigHAPathFlags.path
+	if len(configPath) > 0 {
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateConfigUtils.go:7]: if we are only going to read 1st value in the []string, then why are we getting []string in the function argument.
Would it not be better if in the `runDeployCmd` and `runProvisionInfraCmd` we check the no. of values and pass the first value directly.

File: components/automate-cli/cmd/chef-automate/gather-logs-ha-template.go [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+Usage:
+    chef-automate gather-logs
+Options:
+    -o, --ssh-options OPTIONS          Additional options to pass to ssh
+    -c, --sudo-command SUDO_COMMAND    Alternate command for sudo (default: "sudo")
+    -s, --sudo-options SUDO_OPTIONS    Additional options for remote sudo
+    -r, --remote-log-dir DIRECTORY     Where to temporarily store remote log files (default: "/var/tmp")
+    -l, --local-log-dir DIRECTORY      Where to store the logs locally (default: "/var/tmp")
+    -n, --log-lines NUMBER             How many journalctl lines to capture
+	-h, --help                         Print help
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/gather-logs-ha-template.go:13]: Indentation issues

File: components/automate-cli/cmd/chef-automate/automateHAModeWithoutConfigFile.go [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+func newHaWithoutConfig() *haWithoutConfig {
+	return &haWithoutConfig{}
+}
+
+func (h *haWithoutConfig) doDeployWork(args []string) error {
+	return status.Annotate(errors.New("config.toml file path expected as argument."), status.DeployError)
+}
+
+func (h *haWithoutConfig) doProvisionJob(args []string) error {
+	return status.Annotate(errors.New("config.toml file path expected as argument to run chef-atomate provision-infra"), status.DeployError)
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/automateHAModeWithoutConfigFile.go:19]:
```suggestion
	return status.Annotate(errors.New("config.toml file path expected as argument to run chef-automate provision-infra"), status.DeployError)
```

File: components/automate-cli/cmd/chef-automate/deploy.go [hunk 22]
```diff
@@ -33,6 +33,12 @@ https://www.chef.io/online-master-agreement
 I agree to the Terms of Service and the Master License and Services Agreement
 `
 var errMLSA = "Chef Software Terms of Service and Master License and Services Agreement were not accepted"
+var errProvisonInfra = `Architecture does not match with requested one.
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/deploy.go:36]:
```suggestion
var errProvisonInfra = `Architecture does not match with the requested one. 
```

File: components/automate-cli/cmd/chef-automate/generate-a2harb.go [hunk 23]
```diff
@@ -31,16 +44,24 @@ func readConfigAndWriteToFile() error {
 		if err != nil {
 			return status.Wrap(err, status.ConfigError, "error in unmarshalling config toml file")
 		}
+		errList := validateAwsFields(AwsConfig)
+		if errList != nil && errList.Len() > 0 {
+			return status.Wrap(getSingleErrorFromList(errList), status.ConfigError, "config is invalid.")
+		}
 		finalTemplate := renderSettingsToA2HARBFile(awsA2harbTemplate, AwsConfig)
-		writeToA2HARBFile(finalTemplate, "a2ha.rb")
+		writeToA2HARBFile(finalTemplate, initConfigHabA2HAPathFlag.a2haDirPath+"a2ha.rb")
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/generate-a2harb.go:52]:
```suggestion
		writeToA2HARBFile(finalTemplate, Path.Join(initConfigHabA2HAPathFlag.a2haDirPath, "a2ha.rb"))
```

File: components/automate-cli/cmd/chef-automate/generate-a2harb.go [hunk 24]
```diff
@@ -4,18 +4,31 @@ package main
...
-	writer.Printf("reading configs from toml file")
+func getSingleErrorFromList(l *list.List) error {
+	errorString := ""
+	for e := l.Front(); e != nil; e = e.Next() {
+		errorString = errorString + "\n" + e.Value.(string)
+	}
+	return errors.New(errorString)
+}
+
+func readConfigAndWriteToFile(configPath string) error {
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/generate-a2harb.go:26]: This function also needs to decide on different architectures. We should use deployer and add the respective functions in the struct.

File: components/automate-cli/cmd/chef-automate/provision-infra-ha-template.go [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+	config toml file path
+
+Options:
+    -o, --okta-aws                   Use okta_aws cli to authenticate with AWS
+    -p, --parallelism WORKERS        Number of workers used for the terraform apply (default: 50)
+    -d, --force-deploy               Force terraform deploy even if no changes have been detected. This is used when needing to update the tfstate
+    -y, --accept-changes             Automatically accept and apply terraform changes
+    -n, --skip-changes               Only show what changes are needed, don't apply them
+    -t, --timeout MINUTES            Set timeout value for commands to finish running (default: 10)
+    -h, --help                       print help
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/provision-infra-ha-template.go:20]:
```suggestion
    -h, --help                       Print help
```

File: components/automate-cli/cmd/chef-automate/secrets-ha-template.go [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+Parameters:
+    SUBCOMMAND                       subcommand
+    [ARG] ...                        subcommand arguments
+Subcommands:
+    init                             Generate a new secrets key used to encrypt the secret store.
+    set                              Add a new secrets: for more info use 'secrets set --help'
+    show                             Show existing secrets: for more info use 'secrets show --help'
+    delete                           Delete existing secrets: for more info use 'secrets delete --help'
+Args:
+	automate_admin_password			Set automate admin password
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/secrets-ha-template.go:15]: Fix the indentation

File: components/automate-cli/cmd/chef-automate/workspace-ha-template.go [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+    chef-automate workspace [OPTIONS] SUBCOMMAND [ARG] ...
+Parameters:
+    SUBCOMMAND                       subcommand
+    [ARG] ...                        subcommand arguments
+Subcommands:
+    setup                            Setup a new workspace
+    update                           Update existing workspace
+    export                           Export workspace bundle for use in other environments
+Options:
+    -h, --help                       print help
```
Comment (kalroy) [inline components/automate-cli/cmd/chef-automate/workspace-ha-template.go:14]:
```suggestion
    -h, --help                       Print help
```

File: components/automate-deployment/pkg/target/local_target.go [hunk 28]
```diff
@@ -1520,3 +1520,22 @@ func (t *LocalTarget) IPs() []net.IP {
 func (t *LocalTarget) HabCache() depot.HabCache {
 	return depot.FromLocalCache()
 }
+
+func (t *LocalTarget) InstallAutomateBackendDeployment(ctx context.Context, m manifest.ReleaseManifest) error {
+	//var pkg habpkg.HabPkg = habpkg.New("chef", "automate-backend-deployment")
```
Comment (kalroy) [inline components/automate-deployment/pkg/target/local_target.go:1525]: Remove the comment

File: components/automate-deployment/pkg/target/local_target.go (hunk 1 above)
Comment (kalroy) [inline components/automate-deployment/pkg/target/local_target.go:1531]: As discussed, utilize the `--channel` CLI argument of `deploy` subcommand instead of using hardcoded values.

**Review by kalroy — CHANGES_REQUESTED** [review]
There are unnecessary `else` blocks across the code which needs to be removed.

File: components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go (hunk 18 above)
Comment (kalroy) [reply components/automate-cli/cmd/chef-automate/automateDeploymentFactory.go:3]: Agreed. This file is unnecessary and deployer is enough.

**Review by kalroy — CHANGES_REQUESTED** [review]

**Comment by jayvikramsharma1** [comment]
@vivek-yadav @kalroy comments are resolved now, please have a look

**Review by vivek-yadav — APPROVED** [review]

**Review by kalroy — APPROVED** [review]

**Review by kalroy — APPROVED** [review]
Approved but make sure the pipelines are running fine!! Can see some stable ones failing
